File: src/022/p022.go
package main

import (
	"fmt"
	"io/ioutil"
	"sort"
	"strings"
)

func loadNames() []string {
	file, err := ioutil.ReadFile("p022_names.txt")
	if err != nil {
		fmt.Println("Error reading file.")
	}

	namesString := strings.Replace(string(file), "\"", "", -1)
	names := strings.Split(namesString, ",")
	sort.Strings(names)
	return names
}

func solve() int {
	names := loadNames()
	sum := 0

	// Iterate over names
	for i := range names {
		nameScore := 0
		name := names[i]

		// Chars of name
		for j := range name {
			nameScore += int(name[j]) - 64
		}

		sum += nameScore * (i + 1)
	}

	return sum
}

func main() {
	result := solve()
	fmt.Println(result)
}

File: src/022/p022.ml
open Core_kernel
open Stdio

let replace ~pattern ~with_str =
  Str.global_replace (Str.regexp_string pattern) with_str

let load_names = 
  In_channel.read_all "p022_names.txt"
  |> replace ~pattern:"\"" ~with_str:""
  |> String.split ~on:',' 
  |> List.sort ~compare:(Stdlib.compare)

let sum = List.fold ~f:(+) ~init:(0)

let name_score index name = 
  let char_val c = Char.to_int c - 64 in
  name
  |> Bytes.of_string
  |> Bytes.to_list
  |> List.map ~f:(fun c -> (index + 1) * char_val c)
  |> sum

let solve = 
  load_names
  |> List.mapi ~f:(name_score)
  |> sum

let () =
  printf "%i\n" solve;;
